Enhancement : adding support for Aurora postgres AWS IAM authentication

fabio-rizzo-01 opened the pull request:
Adding ability to connect to AWS Aurora Postgres using IAM authentication (password-less DB connection).

Tests carried out:

1.  Deployed image to EKS cluster with [IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) (IAM Role for Service Account) enabled. Polaris is able to connect to aurora database

2. Deployed image to minikube to validate username/password auth to Postgress DB still works

Both test cases were successful.

Added example configuration in properties file

singhpk234 commented on persistence/relational-jdbc/build.gradle.kts:
```diff
@@ -35,6 +35,9 @@ dependencies {
 
   implementation(libs.smallrye.common.annotation) // @Identifier
   implementation(libs.postgresql)
+  implementation("software.amazon.awssdk:url-connection-client")
```
why do we need url-connection-client specifically ? can this not work with apache http client ?
per this should work - https://docs.quarkiverse.io/quarkus-amazon-services/dev/amazon-rds.html#_configuring_rds_clients

fabio-rizzo-01 replied:
either should work yes, you want me to change it?

dimas-b replied:
My impression from the quarkiverse docs is that the type of HTTP client is a build time decision that is implied by the dependency included in the build (`url-connection-client` vs. `apache-client`). The latter is [already](https://github.com/apache/polaris/blob/4e2facbb4b743b4d54c1d39aaf6fc290960eca7e/runtime/service/build.gradle.kts#L92) included in Polaris.

My preference would be to use the `apache-client` for RDS too and avoid `url-connection-client`.

WDYT? CC: @snazy

fabio-rizzo-01 replied:
I'll update the PR

singhpk234 commented on runtime/defaults/src/main/resources/application.properties:
```diff
@@ -122,9 +122,27 @@ polaris.features."SUPPORTED_CATALOG_CONNECTION_TYPES"=["ICEBERG_REST"]
+quarkus.devservices.enabled=false
+
+
+#AWS IAM postgres example
```
wondering linking this additionally https://docs.quarkiverse.io/quarkus-amazon-services/dev/amazon-rds.html#_configuration_reference would be helpful too

singhpk234 commented on runtime/defaults/src/main/resources/application.properties:
```diff
@@ -122,9 +122,27 @@ polaris.features."SUPPORTED_CATALOG_CONNECTION_TYPES"=["ICEBERG_REST"]
+#quarkus.datasource.jdbc.additional-jdbc-properties.sslmode=require
+#quarkus.datasource.credentials-provider=aws
+
+#quarkus.rds.credentials-provider.aws.use-quarkus-client="true"
+#quarkus.rds.credentials-provider.aws.username=dbusername
```
seems like this is from this section - https://docs.quarkiverse.io/quarkus-amazon-services/dev/amazon-rds.html#_iam_authentication_support_with_quarkus_credential_provider ? 
did you had to change the trust relationships etc and IAM policy etc, would be helpful adding  something about that here too !

wondering if we can move this to https://polaris.apache.org/in-dev/unreleased/metastores/ ? cc @flyrain @dimas-b  thoughts ?

fabio-rizzo-01 replied:
do you want me to add a comment with more details around the policy for the pod? or in the docs? let me know what details you want me to add

dimas-b replied:
Instructions in docs would be nice, but IMHO, it can be a separate PR

singhpk234 reviewed (commented):
Thanks @fabio-rizzo-01 for taking this up !

dimas-b commented on persistence/relational-jdbc/build.gradle.kts:
```diff
@@ -35,6 +35,9 @@ dependencies {
 
   implementation(libs.smallrye.common.annotation) // @Identifier
   implementation(libs.postgresql)
+  implementation("software.amazon.awssdk:apache-client")
+  implementation(platform(libs.quarkus.amazon.services.bom))
+  implementation("io.quarkiverse.amazonservices:quarkus-amazon-rds")
```
I guess these deps are not required for building the JDBC code itself. They probably need to be in `runtime/common`.

The idea is to keep the JDBC persistence lean, but add RDS where drivers would normally be added. WDYT?

(sorry, I missed this in my previous review)

dimas-b replied:
Note: `apache-client` is probably already present in "runtime" modules.

dimas-b commented:
Thanks for your contribution, @fabio-rizzo-01 !

fabio-rizzo-01 commented:
> Thanks a lot @fabio-rizzo-01 for working on it. I'm glad this was figured out. The PR looks good to me overall! My only concern is that users won't be able to figure out how database configure work from a bag of comments from the file `application.properties`. Here are complexity users will face:
> 
> 1. Two options to configure JDBC db. 1. Username/Password 2. AWS IAM AuthN
> 2. Two `application.properties` files(admin/server) to configure them. And they have to be consistent with each other.
> 
> I'd propose to move these configures to a new section under "Relational JDBC" of the page https://polaris.apache.org/in-dev/unreleased/metastores/. With certain description like `there are two options. option 1 ..., option 2 ...`. WDYT?

shall we merge this and I'll create another PR for the docs?

dimas-b commented:
I think doing doc changes in a follow-up PR makes sense :+1:

dimas-b commented on runtime/defaults/src/main/resources/application.properties:
```diff
@@ -122,9 +122,29 @@ polaris.features."SUPPORTED_CATALOG_CONNECTION_TYPES"=["ICEBERG_REST"]
 polaris.persistence.type=in-memory
-# polaris.persistence.type=relational-jdbc
-
+#polaris.persistence.type=relational-jdbc
```
true, @fabio-rizzo-01 : would you mind reverting this line for the sake of proper tranceability should someone need to see when and why it was added?

fabio-rizzo-01 commented on runtime/defaults/src/main/resources/application.properties:
```diff
@@ -122,9 +122,29 @@ polaris.features."SUPPORTED_CATALOG_CONNECTION_TYPES"=["ICEBERG_REST"]
-
+#polaris.persistence.type=relational-jdbc
 polaris.secrets-manager.type=in-memory
+#disable localstack
```
PR updated

fabio-rizzo-01 commented:
> Thanks, @fabio-rizzo-01, for addressing the comment. I'm good with a follow-up doc PR. In that case, I suggest we remove the configuration changes from both `application.properties` files. We add a new jdbc connection option in this PR. We got two options now. Users might confused by which option to choose, how they set it, especially since neither file includes a complete setup for both.

I can't remove all the properties these 2 need to be there: `quarkus.rds.sync-client.type=apache quarkus.devservices.enabled=false`

the app won't run, because it needs to know which client to use to connect otherwise it defaults to the url-connection-client that is not there and for the second one it will try to start up localstack. So I can remove the db config but not these 2.

singhpk234 commented:
> I can't remove all the properties these 2 need to be there: quarkus.rds.sync-client.type=apache quarkus.devservices.enabled=false

This seems reasonable if this is a build time property and all others are runtime properties. my understaing we were concerned mostly from two ways to configure runtime properties for the datasources. 

Orthogonally it would be really nice to add a `CHANGELOG.md` entry for this ! Thank you @fabio-rizzo-01 for working on this one.

fabio-rizzo-01 commented:
> Good point about CHANGELOG (I missed that 😅 ). @fabio-rizzo-01 : could you add a brief notice that the RDS plugin is now enabled and allows connecting to Aurora DB? I do not think it needs to be too verbose since doc changes are coming later.

properties removed and updated changelog file

flyrain reviewed (approved):
+1 Thanks a lot @fabio-rizzo-01 for adding this!
